Argument for Expected response code Fixes (9)

Safouene1 opened the pull request:
# Description

Adds an **optional** regex argument `-regex=<pattern>` for response codes instead of checking success codes only.
if no value is assigned to -regex , wait-for checks success codes as usual. 

Fixes #9

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

I've integrated the defaultRegex so it won't break in all test cases and Added 2 new test cases: 
- defaultRegexCanBeSet : Checks if regex argument can be set
-failsIfRegexInvalid: Checks if command fail when provided with bad regular expressions. 

- [x] Running existing tests
- [x] Created new tests

# Checklist:

- [x] My code has been linted (`make lint`)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes (`make test lint build acceptance-test`)
- [x] I have rebased my branch to include the latest changes from `master`

ghost commented:
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=20&src=https%3A%2F%2Fgithub.com%2Fdnnrly%2Fwait-for">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=20&src=https%3A%2F%2Fgithub.com%2Fdnnrly%2Fwait-for)
</details>

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/dnnrly/wait-for/20/cda8ab07/eca1ab1f95519a1446fc086a2d90ea723895c138.svg)](https://app.codesee.io/r/reviews?pr=20&src=https%3A%2F%2Fgithub.com%2Fdnnrly%2Fwait-for)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

dnnrly commented on cmd/wait-for/main.go:
```diff
@@ -18,11 +18,13 @@ func main() {
 	flag.StringVar(&httpTimeoutParam, "http_timeout", httpTimeoutParam, "timeout for requests made by a http client")
 	flag.StringVar(&configFile, "config", "", "configuration file to use")
 	flag.BoolVar(&quiet, "quiet", false, "reduce output to the minimum")
+	flag.StringVar(&regexStatus, "regex", "", "Use regex to match the expected result in HTTP status codes")
```
```suggestion
	flag.StringVar(&statusPattern, "status", "200", "A golang regex that represents the desired HTTP status response code")
```

You'll probably have to rename a few instances of variable used here.

Safouene1 replied:
I will rename it statusPatternParam , Do you approve?

dnnrly replied:
Yes, you're right. That follows the convention.

dnnrly commented on config.go:
```diff
@@ -15,6 +15,9 @@ const DefaultTimeout = time.Second * 5
 const DefaultHTTPClientTimeout = time.Second
 
+// DefaultRegexStatus a default value for the Regex pattern to match in the expected result
+const DefaultRegexStatus = ""
```
```suggestion
const DefaultDStatusPattern = ""
```

dnnrly commented on config.go:
```diff
@@ -25,13 +28,16 @@ type TargetConfig struct {
 	DefaultTimeout           time.Duration `yaml:"default-timeout"`
 	Targets                  map[string]TargetConfig
 	DefaultHTTPClientTimeout time.Duration `yaml:"default-http-client-timeout"`
+	DefaultRegexStatus       string        `yaml:"default-http-client-regex"`
```
```suggestion
	StatusPattern       string        `yaml:"default-http-client-status-pattern"`
```

dnnrly commented on config.go:
```diff
@@ -105,7 +119,7 @@ func (c *Config) AddFromString(t string) error {
 	}
 
-	return errors.New("unable to understand target " + t)
+	return errors.New("unable to understand target  " + t)
```
```suggestion
	return errors.New("unable to understand target " + t)
```

dnnrly commented on waitfor.go:
```diff
@@ -154,9 +156,20 @@ func HTTPWaiter(name string, target *TargetConfig) error {
-
-	if !isSuccess(resp.StatusCode) {
-		return fmt.Errorf("got %d from %s", resp.StatusCode, name)
+	if target.RegexStatus != "" {
```
I think if we use the default of `200` in the parameter, we can remove the code paths that deal with an empty pattern. It'll also boost the code coverage.

Safouene1 replied:
That's brilliant , I don't know why I didn't think about it !

dnnrly commented on waitfor.go:
```diff
@@ -58,7 +60,7 @@ func WaitOn(config *Config, logger Logger, targets []string, waiters map[string]
 }
 
-func OpenConfig(configFile, defaultTimeout, defaultHTTPTimeout string, fs afero.Fs) (*Config, error) {
+func OpenConfig(configFile, defaultTimeout, defaultHTTPTimeout string, fs afero.Fs, defaultRegexPattern string) (*Config, error) {
```
```suggestion
func OpenConfig(configFile, defaultTimeout, defaultHTTPTimeout, defaultStatusPattern string, fs afero.Fs) (*Config, error) {
```

dnnrly commented on waitfor_test.go:
```diff
@@ -83,11 +83,19 @@ func TestOpenConfig_defaultHTTPTimeoutCanBeSet(t *testing.T) {
 	assert.NotNil(t, config)
 	assert.Equal(t, time.Second*20, config.DefaultHTTPClientTimeout)
 }
+func TestOpenConfig_defaultRegexCanBeSet(t *testing.T) {
```
```suggestion

func TestOpenConfig_defaultRegexCanBeSet(t *testing.T) {
```

dnnrly commented on waitfor_test.go:
```diff
@@ -137,7 +145,22 @@ func TestWaitOnSingleTarget_succeedsAfterWaiting(t *testing.T) {
 	assert.Contains(t, logs, "error while waiting for name: there was an error")
 	assert.Contains(t, logs, "finished waiting for name")
 }
+func TestWaitOnSingleTarget_failsIfRegexInvalid(t *testing.T) {
```
```suggestion

func TestWaitOnSingleTarget_failsIfRegexInvalid(t *testing.T) {
```

dnnrly commented:
Honestly, I like this PR a lot! The changes I suggested a important but basically cosmetic.

There are only 2 things that I would like to see in addition:

1. A bit more code coverage - codecov.io is blocking the merge
2. A couple of acceptance tests to prove the plumbing of the CLI parameter and also the yaml config

If you're not familiar with Cucumber/Godog tests then I'm happy to push a couple of commits, if you'd like.


Thank you so much for your contribution.

Safouene1 commented:
Thank you ! I will do the required changes .
I Don't know Godog but I will have a look into it and add the required tests.

dnnrly commented on waitfor.go:
```diff
@@ -58,7 +60,7 @@ func WaitOn(config *Config, logger Logger, targets []string, waiters map[string]
 }
 
-func OpenConfig(configFile, defaultTimeout, defaultHTTPTimeout string, fs afero.Fs) (*Config, error) {
+func OpenConfig(configFile, defaultTimeout, defaultHTTPTimeout string, fs afero.Fs, defaultStatusPattern string) (*Config, error) {
```
```suggestion
func OpenConfig(configFile, defaultTimeout, defaultHTTPTimeout, defaultStatusPattern string, fs afero.Fs) (*Config, error) {
```

dnnrly commented on waitfor.go:
```diff
@@ -179,6 +180,26 @@ func GRPCWaiter(name string, target *TargetConfig) error {
 	return nil
 }
 
+// checkStatus checks if the given HTTP status code matches the pattern provided in the target configuration.
+func checkStatus(targetPattern string, code int) error {
+	if targetPattern != "200" {
+		// Safely compile and initialize  the regular expression pattern and verify if it's valid
+		pattern, err := regexp.Compile(targetPattern)
+		if err != nil {
+			return fmt.Errorf("invalid Regular Expression %v", err)
+		}
+		if !pattern.MatchString(strconv.Itoa(code)) {
+			return fmt.Errorf("%d status Code and %s regex didn't match", code, pattern.String())
+		}
+		// if the target is set to default "200", check if the status code is a successful status code
+	} else {
+		if !isSuccess(code) {
+			return fmt.Errorf("got %d ", code)
+		}
+	}
+	return nil
+}
```
I've been looking at this a little more. I wonder if we make the default pattern `^2..$` then we can remove the `else` code block and the `isSuccess` method entirely.

What do you think?

Safouene1 replied:
Yes I agree it would be much better that way.

Safouene1 commented:
Updated the code with the  the default pattern ^2..$ and removed the Else block with the isSuccess function (and it's tests).
I will have a look into Godog and the failing acceptance tests.

dnnrly commented:
Excellent, thanks. I think this is all heading in the right direction. Once those acceptance tests are sorted then I think we're good.

dnnrly commented on test/features/cli.feature:
```diff
@@ -76,6 +76,28 @@ Feature: Configuration from CLI
     And the time taken is less than "5s"
     And the output contains "timed out waiting for http://non-existent/health"
 
+  Scenario: Status is configurable
+    Given I have an HTTP server running on port 80 that responds with 400
+    When I run wait-for with parameters "-status 400 http://localhost/health"
+    Then I can see that an HTTP request was made for "localhost GET /health"
+    And the output contains "finished waiting for http://localhost/health"
+    And wait-for exits without error
+
+  Scenario: Fails when status is not a valid regex
+    Given I have an HTTP server running on port 80 that responds with 200
+    When I run wait-for with parameters "-status [4-0]{3} http://localhost/health"
+    Then I can see that an HTTP request was made for "localhost GET /health"
+    And the output contains " error while waiting for http://localhost/health:  invalid Regular Expression"
+    And wait-for exits with an error
+
+  Scenario: Fails when status doesn't match response
+    Given I have an HTTP server running on port 80 that responds with 200
+    When I run wait-for with parameters "-status 500 http://localhost/health"
+    Then I can see that an HTTP request was made for "localhost GET /health"
+    And the output contains " error while waiting for http://localhost/health:  200 status Code and 500 regex didn't match"
+    And wait-for exits with an error
```
I like these tests!

Safouene1 replied:
Thanks ! Godog is pretty fun .
I learned new fun things doing this !

dnnrly commented:
Many thanks for the contribution @Safouene1. You worked hard on this and it shows. As is tradition in my repos, if you do twitter then I will gladly follow you.
